File: Cantonese-Mandarin sound correspondences/rimes.jl
include("../helpers.jl")

using Unicode

struct PinyinSyllable
   initial::Union{String, Nothing}
   pseudo_initial::Union{String, Nothing}
   final::String
   pseudo_final::String
   original::String
end

struct JyutpingSyllable
   initial::Union{String, Nothing}
   pseudo_initial::Union{String, Nothing}
   final::String
   pseudo_final::String
   final_plosive::Union{String, Nothing}
   original::String
end

PINYIN_SYLLABIC_FRICATIVES = Set([
   "zi",
   "ci",
   "si",
   "zhi",
   "chi",
   "shi",
   "ri"
])

PINYIN_INITIALS = [
   "zh",
   "ch",
   "sh",
   "b",
   "p",
   "m",
   "f",
   "d",
   "t",
   "n",
   "z",
   "c",
   "s",
   "l",
   "r",
   "j",
   "q",
   "x",
   "g",
   "k",
   "h"
]

PINYIN_FINALS = Set([
   "e",
   "a",
   "ei",
   "ai",
   "ou",
   "ao",
   "en",
   "an",
   "ong",
   "eng",
   "ang",
   "er",
   "yi",
   "ye",
   "ya",
   "you",
   "yao",
   "yin",
   "yan",
   "ying",
   "yong",
   "yang",
   "wu",
   "wo",
   "wa",
   "wei",
   "wai",
   "wen",
   "wan",
   "weng",
   "wang",
   "yo",
   "yu",
   "yue",
   "yun",
   "yuan"
])

PINYIN_NONNULL_INITIAL_FINALS = Dict([
   "i" => "yi",
   "ie" => "ye",
   "ia" => "ya",
   "iu" => "you",
   "iao" => "yao",
   "in" => "yin",
   "ian" => "yan",
   "ing" => "ying",
   "iong" => "yong",
   "iang" => "yang",
   "u" => "wu",
   "o" => "wo",
   "uo" => "wo",
   "ua" => "wa",
   "ui" => "wei",
   "uai" => "wai",
   "un" => "wen",
   "uan" => "wan",
   "uang" => "wang",
   "u:" => "yu",
   "u:e" => "yue",
   "u:n" => "yun",
   "u:an" => "yuan"
])

function parse_pinyin(pinyin::AbstractString)
   syllable = pinyin[1:end-1]

   if syllable in PINYIN_SYLLABIC_FRICATIVES
      return PinyinSyllable(syllable[1:end-1], syllable[1:end-1], "_", "_", syllable)
   end

   initial_index = findfirst(i -> startswith(syllable, i), PINYIN_INITIALS)
   initial = initial_index === nothing ? nothing : PINYIN_INITIALS[initial_index]
   
   final = begin
      remaining = if initial === nothing
         syllable
      else
         syllable[length(initial)+1:end]
      end

      if startswith(remaining, "u") && initial in ["j", "q", "x"]
         PINYIN_NONNULL_INITIAL_FINALS[replace(remaining, "u" => "u:")]
      elseif haskey(PINYIN_NONNULL_INITIAL_FINALS, remaining)
         PINYIN_NONNULL_INITIAL_FINALS[remaining]
      elseif remaining in PINYIN_FINALS
         remaining
      else
         nothing
      end
   end

   if final === nothing
      error("Invalid pinyin: $pinyin")
   end

   pseudo_initial = if initial === nothing && (startswith(final, "y") || startswith(final, "w"))
      final[1:1]
   elseif (initial === "g" || initial === "k") && startswith(final, "w")
      initial * "w"
   else
      initial
   end

   pseudo_final = if pseudo_initial === "kw" || pseudo_initial === "gw"
      final[2:end]
   elseif initial === nothing && (startswith(final, "w") || (startswith(final, "y") && !startswith(final, "yu")))
      final[2:end]
   else
      final
   end

   PinyinSyllable(initial, pseudo_initial, final, pseudo_final, syllable)
end

JYUTPING_SYLLABIC_NASALS = ["m", "ng"]

JYUTPING_INITIALS = [
   "ng",
   "gw",
   "kw",
   "b",
   "p",
   "m",
   "f",
   "d",
   "t",
   "n",
   "s",
   "l",
   "z",
   "c",
   "j",
   "g",
   "k",
   "w",
   "h"
]

JYUTPING_FINALS = [
   "aa",
   "aai",
   "aau",
   "aam",
   "aan",
   "aang",
   "aap",
   "aat",
   "aak",
   "a",
   "ai",
   "au",
   "am",
   "an",
   "ang",
   "ap",
   "at",
   "ak",
   "e",
   "ei",
   "eu",
   "em",
   "eng",
   "ep",
   "ek",
   "i",
   "iu",
   "im",
   "in",
   "ing",
   "ip",
   "it",
   "ik",
   "o",
   "oi",
   "ou",
   "on",
   "ong",
   "ot",
   "ok",
   "u",
   "ui",
   "un",
   "ung",
   "ut",
   "uk",
   "eoi",
   "eon",
   "eot",
   "oe",
   "oeng",
   "oet",
   "oek",
   "yu",
   "yun",
   "yun",
   "yut"
]

function parse_jyutping(jyutping::String)
   syllable = jyutping[1:end-1]
   
   if syllable in JYUTPING_SYLLABIC_NASALS
      return JyutpingSyllable(nothing, nothing, syllable, syllable, nothing, syllable)
   end

   initial_index = findfirst(i -> startswith(syllable, i), JYUTPING_INITIALS)
   initial = initial_index === nothing ? nothing : JYUTPING_INITIALS[initial_index]

   (final, final_plosive) = begin
      remaining = syllable[length(default(initial, ""))+1:end]
      final_plosive = if any(p -> endswith(remaining, p), ["p", "t", "k"])
         remaining[end:end]
      else
         nothing
      end

      if remaining in JYUTPING_FINALS
         (remaining[1:end-length(default(final_plosive, ""))], final_plosive)
      else
         nothing
      end
   end

   pseudo_initial = if initial === "kw" || initial === "gw"
      initial[1:1]
   else
      initial
   end

   if final === nothing
      error("Invalid jyutping: $jyutping")
   end

   pseudo_final = if initial === "w" || (initial === "j" && !startswith(final, "yu"))
      initial * final
   elseif initial === "kw" || initial === "gw"
      "w" * final
   else
      final
   end

   JyutpingSyllable(initial, pseudo_initial, final, pseudo_final, final_plosive, syllable)
end

function syllable_mapping(triples)
   monophones = filter(t -> length(t[2]) == 1, triples)
   syllable_triples = map(monophones) do t
      (t[1], parse_pinyin(t[2] |> only), parse_jyutping(t[3]))
   end

   mc_initial_mapping = Dict()
   mc_final_mapping = Dict()

   cm_initial_mapping = Dict()
   cm_final_mapping = Dict()

   for (c, pinyin, jyutping) in syllable_triples
      push!(get!(get!(mc_initial_mapping, pinyin.pseudo_initial, Dict()), jyutping.initial, Set()), c)
      push!(get!(get!(mc_final_mapping, pinyin.pseudo_final, Dict()), jyutping.final, Set()), c)

      push!(get!(get!(cm_initial_mapping, jyutping.pseudo_initial, Dict()), pinyin.initial, Set()), c)
      push!(get!(get!(cm_final_mapping, jyutping.pseudo_final, Dict()), pinyin.final, Set()), c)
   end

   Dict(
      "mc" => (mc_initial_mapping, mc_final_mapping),
      "cm" => (cm_initial_mapping, cm_final_mapping)
   )
end

File: Cantonese-Mandarin sound correspondences/rimes.ts
import { groupBy, mapEntries } from "https://deno.land/std@0.116.0/collections/mod.ts";
import { PJChar } from "./data/index.ts";

const PINYIN_SYLLABIC_FRICATIVES = [
   "zi",
   "ci",
   "si",
   "zhi",
   "chi",
   "shi",
   "ri",
];

const PINYIN_INITIALS = [
   "zh",
   "ch",
   "sh",
   "b",
   "p",
   "m",
   "f",
   "d",
   "t",
   "n",
   "z",
   "c",
   "s",
   "l",
   "r",
   "j",
   "q",
   "x",
   "g",
   "k",
   "h",
];

const PINYIN_FINALS = [
   "e",
   "a",
   "ei",
   "ai",
   "ou",
   "ao",
   "en",
   "an",
   "ong",
   "eng",
   "ang",
   "er",
   "yi",
   "ye",
   "ya",
   "you",
   "yao",
   "yin",
   "yan",
   "ying",
   "yong",
   "yang",
   "wu",
   "wo",
   "wa",
   "wei",
   "wai",
   "wen",
   "wan",
   "weng",
   "wang",
   "yu",
   "yue",
   "yun",
   "yuan",
];

const PINYIN_NONNULL_INITIAL_FINALS: Record<string, string> = {
   "i": "yi",
   "ie": "ye",
   "ia": "ya",
   "iu": "you",
   "iao": "yao",
   "in": "yin",
   "ian": "yan",
   "ing": "ying",
   "iong": "yong",
   "iang": "yang",
   "u": "wu",
   "o": "wo",
   "uo": "wo",
   "ua": "wa",
   "ui": "wei",
   "uai": "wai",
   "un": "wen",
   "uan": "wan",
   "uang": "wang",
   "ü": "yu",
   "üe": "yue",
   "ün": "yun",
   "üan": "yuan",
};

export interface PinyinSyllable {
   initial?: string;
   final: string;
   full: string;
}

function parsePinyin(pinyin: string): PinyinSyllable {
   const normalized = pinyin.normalize("NFD").split("").filter((l) => l.match(/[a-z\u0308]/g))
      .join("")
      .normalize("NFC");

   if (PINYIN_SYLLABIC_FRICATIVES.indexOf(normalized) >= 0) {
      return {
         initial: normalized.substring(0, normalized.length - 1),
         final: "_",
         full: normalized,
      };
   }

   const initial = PINYIN_INITIALS.find((i) => normalized.startsWith(i));

   const final = (() => {
      const remaining = initial == null ? normalized : normalized.substring(initial.length);

      if (remaining.startsWith("u") && ["j", "q", "x"].indexOf(initial ?? "") >= 0) {
         return PINYIN_NONNULL_INITIAL_FINALS[remaining.replace("u", "ü")];
      } else if (Object.hasOwn(PINYIN_NONNULL_INITIAL_FINALS, remaining)) {
         return PINYIN_NONNULL_INITIAL_FINALS[remaining];
      } else if (PINYIN_FINALS.indexOf(remaining) >= 0) {
         return remaining;
      }

      return null;
   })();

   if (final == null) {
      throw new Error(`Invalid pinyin: ${pinyin}`);
   }

   const pseudoInitial = (() => {
      if ((initial == "k" || initial == "g") && (final.startsWith("w") || final.startsWith("u"))) {
         return initial + "w";
      }

      if (initial == null && (final.startsWith("w") || final.startsWith("y"))) {
         return final.substr(0, final.startsWith("yu") ? 2 : 1);
      }
   })();

   return {
      initial: pseudoInitial ?? initial,
      final,
      full: normalized,
   };
}

export interface JyutpingSyllable {
   initial?: string;
   final: string;
   finalPlosive?: string;
   full: string;
}

const JYUTPING_SYLLABIC_NASALS = ["m", "ng"];

const JYUTPING_INITIALS = [
   "ng",
   "gw",
   "kw",
   "b",
   "p",
   "m",
   "f",
   "d",
   "t",
   "n",
   "s",
   "l",
   "z",
   "c",
   "j",
   "g",
   "k",
   "w",
   "h",
];

const JYUTPING_FINALS = [
   "aa",
   "aai",
   "aau",
   "aam",
   "aan",
   "aang",
   "aap",
   "aat",
   "aak",
   "a",
   "ai",
   "au",
   "am",
   "an",
   "ang",
   "ap",
   "at",
   "ak",
   "e",
   "ei",
   "eu",
   "em",
   "eng",
   "ep",
   "ek",
   "i",
   "iu",
   "im",
   "in",
   "ing",
   "ip",
   "it",
   "ik",
   "o",
   "oi",
   "ou",
   "on",
   "ong",
   "ot",
   "ok",
   "u",
   "ui",
   "un",
   "ung",
   "ut",
   "uk",
   "eoi",
   "eon",
   "eot",
   "oe",
   "oeng",
   "oet",
   "oek",
   "yu",
   "yun",
   "yun",
   "yut",
];

function parseJyutping(jyutping: string): JyutpingSyllable {
   const syllable = jyutping.split("").filter((c) => c.toUpperCase() != c.toLowerCase()).join("");
   if (JYUTPING_SYLLABIC_NASALS.indexOf(syllable) >= 0) {
      return {
         initial: syllable,
         final: syllable,
         full: syllable,
      };
   }

   const initial = JYUTPING_INITIALS.find((i) => syllable.startsWith(i));

   const { final, finalPlosive } = (() => {
      const remaining = syllable.substring(initial?.length ?? 0);
      const finalPlosive = ["p", "t", "k"].find((p) => remaining.endsWith(p));
      if (JYUTPING_FINALS.indexOf(remaining) >= 0) {
         return {
            final: remaining.substring(0, remaining.length - (finalPlosive?.length ?? 0)),
            finalPlosive,
         };
      }
      return {};
   })();

   const finalWithOnset = initial === "j" || initial === "w" ? initial + final : null;

   if (final == null) {
      throw new Error(`Invalid jyutping: ${jyutping}`);
   } else {
      return {
         initial,
         final: finalWithOnset ?? final,
         finalPlosive,
         full: syllable,
      };
   }
}

type PartMapping = Record<string, Record<string, string[]>>;

export interface SyllableMapping {
   initialMapping: PartMapping;
   finalMapping: PartMapping;
}

export interface BidirectionalSyllableMapping {
   mc: SyllableMapping;
   cm: SyllableMapping;
}

interface ParsedCharacter {
   character: string;
   pinyin: PinyinSyllable;
   jyutping: JyutpingSyllable;
}

function groupPart(
   characters: ParsedCharacter[],
   key: "pinyin" | "jyutping",
   part: "initial" | "final",
): PartMapping {
   const outer = groupBy(characters, (c) => c[key][part] ?? "");
   return mapEntries(outer, (entry) => {
      const [syllablePart, characters] = entry;
      const groups = groupBy(
         characters,
         (c) => c[key === "pinyin" ? "jyutping" : "pinyin"][part] ?? "",
      );
      const inner = mapEntries(groups, (entry) => [entry[0], entry[1].map((c) => c.character)]);
      return [syllablePart, inner];
   });
}

export function mapSyllables(characters: PJChar[]): BidirectionalSyllableMapping {
   const monophones = characters.filter((c) => c.pinyin.length === 1);
   const syllableCharacters = monophones.map((c): ParsedCharacter => ({
      character: c.character,
      pinyin: parsePinyin(c.pinyin[0]),
      jyutping: parseJyutping(c.jyutping),
   }));

   return {
      mc: {
         initialMapping: groupPart(syllableCharacters, "pinyin", "initial"),
         finalMapping: groupPart(syllableCharacters, "pinyin", "final"),
      },
      cm: {
         initialMapping: groupPart(syllableCharacters, "jyutping", "initial"),
         finalMapping: groupPart(syllableCharacters, "jyutping", "final"),
      },
   };
}
